### PR TITLE
RFC: use nvim-treesitter's highlight group

### PR DESCRIPTION
**Pull request by fannheyward**
- https://github.com/neoclide/coc.nvim/issues/3345
- https://github.com/neoclide/coc.nvim/issues/3323

what've done:

1. use TS* hi group
2. add `coc.preferences.semanticTokensFiletypes` to enable semanticTokens for special language
3. remove `g:coc_default_semantic_highlight_groups` and `CocSem_*` hi groups

### CODE REVIEW

**Comment by oblitum** [comment]
Will this require to explicitly enable the languages for semantic tokens to work? My original suggestion was to have semantic tokens by default supported, but that when the user wants to, if it wants so, it could configure specific languages, or also, have it blacklisted for just a few.

**Comment by fannheyward** [comment]
> Will this require to explicitly enable the languages for semantic tokens to work

I prefer to disable semanticTokens by default, because only clangd/rust-analzyer/gopls supports this by now, users can enable it by themself. Another reason is semanticTokens is not stable enough by now.

**Comment by oblitum** [comment]
@fannheyward okay, I commented that because it could be annoying to have to list the languages, but I saw after that there's glob support.

**Comment by oblitum** [comment]
At that, it should be documented in the added docs that glob pattern can be used to have it enabled for any server that supports it.

**Comment by fannheyward** [comment]
>  be documented in the added docs that glob pattern

Done.

File: doc/coc.txt [hunk 1]
```diff
@@ -2821,7 +2821,8 @@ Semantic highlights~				*coc-semantic-highlights*
   coc.nvim can use semantic tokens from language server to add additional color
   information that depends on language specific symbol information. The semantic
   tokens highlighting is disabled by default, you can enable it by add filetype
```
Comment (oblitum) [inline doc/coc.txt:2823]: s/add filetype/adding a filetype/

File: data/schema.json [hunk 2]
```diff
@@ -1236,10 +1236,13 @@
       "description": "Enable color highlight if language server support it.",
       "default": true
     },
-    "coc.preferences.semanticTokensHighlights": {
-      "type": "boolean",
-      "description": "Enable semanticTokens highlight if language server support it.",
-      "default": true
+    "coc.preferences.semanticTokensFiletypes": {
+      "type":"array",
+      "description": "Filetypes that enable semanticTokens highlighting",
```
Comment (oblitum) [inline data/schema.json:1241]: "Filetypes that enable semanticTokens highlighting or [\"*\"] for any filetype"

**Comment by oblitum** [comment]
Seems worth noticing that removing the simple default groups provided by `g:coc_default_semantic_highlight_groups` means to remove vim support for this kind of thing.

**Comment by fannheyward** [comment]
>  means to remove vim support for this kind of thing

No, if the color scheme has TS* hi group defined for vim, for example https://github.com/sainnhe/sonokai/blob/61b796d94e7b0110f1bdf321cc1d19a841697a42/colors/sonokai.vim#L331, semanticTokens highlight supports vim too.

**Comment by oblitum** [comment]
@fannheyward wow, seriously, that's really nice!

**Comment by oblitum** [comment]
Glob pattern isn't working anymore, dunno what happened.

**Comment by darzok0914** [comment]
Hi, I'm currently using the CocSem_* highlight groups and I think they are great and a good complement of the TS* groups. Typically, C++ is quite hard to parse and understand for TreeSitter and clangd can provide more information. What I'm saying is that the highlight groups from Tree sitter and the one provided from lsp are not a perfect 1 to 1 mapping and we can have a better experience by combining them instead of trying to map them as it would result in a loss of information.

**Comment by fannheyward** [comment]
cc @oblitum `tokenModifiers` has been handled, something like `TSDeprecated` will be highlighted.

**Comment by oblitum** [comment]
> cc @oblitum `tokenModifiers` has been handled, something like `TSDeprecated` will be highlighted.

Nice. I tried it with Go (I had to `:hi link TSDeprecated CocDeprecatedHighlight` manually) but got the highlighting only at definition, never at call site:

img-2021-12-15-101132

All the function calls inside main are for deprecated functions.

> The problem is, hi group like `TSStatic/TSDeprecated/TSReadonly` are not defined by nvim-treesitter, we should provide these hi group, or user needed to define by themself.

Is it good idea to fallback highlighting (to `CocDeprecatedHighlight`) as commented in https://github.com/neoclide/coc.nvim/pull/3188#issuecomment-974490202?

**Comment by fannheyward** [comment]
```go
package main

// Deprecated, use newFoof
func fooFunc() bool {
	return false
}

// Deprecated
const badPi = 3.141

func main() {
	fooFunc()
}
```

semanticTokens:

```
[
  { hlGroup: 'TSKeyword', lnum: 0, colStart: 0, colEnd: 7 },
  { hlGroup: 'TSNamespace', lnum: 0, colStart: 8, colEnd: 12 },
  { hlGroup: 'TSComment', lnum: 2, colStart: 0, colEnd: 26 },
  { hlGroup: 'TSKeyword', lnum: 3, colStart: 0, colEnd: 4 },
  { hlGroup: 'TSFunction', lnum: 3, colStart: 5, colEnd: 12 },
  { hlGroup: 'TSDefinition', lnum: 3, colStart: 5, colEnd: 12 },
  { hlGroup: 'TSDeprecated', lnum: 3, colStart: 5, colEnd: 12 },
  { hlGroup: 'TSType', lnum: 3, colStart: 15, colEnd: 19 },
  { hlGroup: 'TSDefaultLibrary', lnum: 3, colStart: 15, colEnd: 19 },
  { hlGroup: 'TSKeyword', lnum: 4, colStart: 1, colEnd: 7 },
  { hlGroup: 'TSVariable', lnum: 4, colStart: 8, colEnd: 13 },
  { hlGroup: 'TSReadonly', lnum: 4, colStart: 8, colEnd: 13 },
  { hlGroup: 'TSComment', lnum: 7, colStart: 0, colEnd: 13 },
  { hlGroup: 'TSKeyword', lnum: 8, colStart: 0, colEnd: 5 },
  { hlGroup: 'TSVariable', lnum: 8, colStart: 6, colEnd: 11 },
  { hlGroup: 'TSDefinition', lnum: 8, colStart: 6, colEnd: 11 },
  { hlGroup: 'TSReadonly', lnum: 8, colStart: 6, colEnd: 11 },
  { hlGroup: 'TSDeprecated', lnum: 8, colStart: 6, colEnd: 11 },
  { hlGroup: 'TSNumber', lnum: 8, colStart: 14, colEnd: 19 },
  { hlGroup: 'TSKeyword', lnum: 10, colStart: 0, colEnd: 4 },
  { hlGroup: 'TSFunction', lnum: 10, colStart: 5, colEnd: 9 },
  { hlGroup: 'TSDefinition', lnum: 10, colStart: 5, colEnd: 9 },
  { hlGroup: 'TSFunction', lnum: 11, colStart: 1, colEnd: 8 }
]
```

No `TSDeprecated` on call sites.

**Comment by oblitum** [comment]
Hmm, okay, thanks for checking this, it's their then on that.

**Comment by jackhub** [comment]
I am wondering does this will fix the problem I mentioned for clangd.
https://github.com/clangd/clangd/issues/985

**Comment by oblitum** [comment]
@jackhub just give it a try with a tree-sitter colorscheme.

**Comment by oblitum** [comment]
>  chemzqm force-pushed the feat/use-ts-hi-group branch from 788d4ee to 759cd81

😢 😕

**Comment by jackhub** [comment]
> @jackhub just give it a try with a tree-sitter colorscheme.

I use the latest coc.nvim, how to configure it to have a try with tree sitter color scheme ?

**Comment by oblitum** [comment]
@jackhub give it a try by pulling and building the branch of this pull request.

**Comment by jackhub** [comment]
> @jackhub give it a try by pulling and building the branch of this pull request. [This is my configuration for neovim, for example](https://codeberg.org/oblitum/dotfiles/src/commit/b5161cbf6b17b03a3c1e678d45efaca9f9400817/.config/nvim/init.lua#L851-L852).

Cool, I have tested with this, it works.
image

I have a suggestion here. I want to highlight `DefaultLibaray` type, property, function, namespace different.
Here we just have a `TSDefaultLibaray` for types.

image
Like what Xcode does.

Thanks for your work.

**Comment by jackhub** [comment]
> > @jackhub give it a try by pulling and building the branch of this pull request. [This is my configuration for neovim, for example](https://codeberg.org/oblitum/dotfiles/src/commit/b5161cbf6b17b03a3c1e678d45efaca9f9400817/.config/nvim/init.lua#L851-L852).
> 
> Cool, I have tested with this, it works. image
> 
> I have a suggestion here. I want to highlight `DefaultLibaray` type, property, function, namespace differently. Here we just have a `TSDefaultLibaray` for types.
> 
> Also for declarations, I want to highlight different types of 'Declaration`. Like this, I want to keep macro declaration to be orange color. image
> 
> Can we have to method to custom highlights for the all combination of token types and modifiers?
> 
> Maybe you can highlight all token types as default, ignore modifiers, users can define more precise highlight with modifier, like `TSTypeDeclaration`.
> 
> image
> 
> Like what Xcode does.
> Thanks for your work.

I tried myself to deal with `DefaultLibrary` and `Declaration` modifiers, make it exactly same to Xcode highlight, but it's too customization, maybe coc.nvim can provide some better strategy for this.

`      var modifiers = legend.tokenModifiers.filter((_, m) => tokens[i + 4] & (1 << m)).map(m => SEMANTIC_HLGROUP_PREFIX + m)
      var group = SEMANTIC_HLGROUP_PREFIX + legend.tokenTypes[tokenType]
      if (modifiers.indexOf("TSDefaultLibrary") > -1) {
        if (legend.tokenTypes[tokenType] == "Class" ||
          legend.tokenTypes[tokenType] == "Interface" ||
          legend.tokenTypes[tokenType] == "Enum" ||
          legend.tokenTypes[tokenType] == "Type" ||
          legend.tokenTypes[tokenType] == "Namespace") {
          group = "TSTypeDefaultLibrary"
        } else {
          group = "TSOtherDefaultLibrary"
        }
        modifiers = []
      } else if (modifiers.indexOf("TSDeclaration") > -1) {
        if (legend.tokenTypes[tokenType] == "Class" ||
          legend.tokenTypes[tokenType] == "Interface" ||
          legend.tokenTypes[tokenType] == "Enum" ||
          legend.tokenTypes[tokenType] == "Type" ||
          legend.tokenTypes[tokenType] == "Namespace") {
          group = "TSTypeDeclaration"
          modifiers = []
        } else if (legend.tokenTypes[tokenType] == "Macro") {
        } else if (legend.tokenTypes[tokenType] == "Property" ||
          legend.tokenTypes[tokenType] == "Variable" ||
          legend.tokenTypes[tokenType] == "Parameter") {
          group = "TSVariable"
        } else {
          group = "TSOtherDeclaration"
        }
        modifiers = []
      } else {
        modifiers = []
      }
`
image

File: data/schema.json [hunk 3]
```diff
@@ -1305,10 +1305,13 @@
       "default": "SNIP",
       "description": "Text shown in statusline to indicate snippet session is activated."
     },
-    "coc.preferences.semanticTokensHighlights": {
-      "type": "boolean",
-      "description": "Enable semanticTokens highlight if language server support it.",
-      "default": true
+    "coc.preferences.semanticTokensFiletypes": {
```
Comment (chemzqm) [inline data/schema.json:1308]: My suggestion is use `semanticTokens.filetypes` instead and we will have more configuration for semantic highlights.

File: doc/coc.txt [hunk 4]
```diff
@@ -2917,39 +2915,18 @@ Others~
 vim doesn't support change highlight group of cursorline inside popup.
 *CocSelectedRange* for highlight ranges of outgoing calls.
 
-Semantic highlights~
-
-CocSem_*						*CocSem*
```
Comment (chemzqm) [inline doc/coc.txt:2922]: My suggestion is still use `CocSem` as prefix of highlight group and link to `TS` highlight groups when possible, so they're decoupled.

**Review by chemzqm — CHANGES_REQUESTED** [review]

**Comment by oblitum** [comment]
never mind, it's working, false alarm.

**Comment by chemzqm** [comment]
Need create link to `TS*` highlight groups at https://github.com/neoclide/coc.nvim/blob/master/plugin/coc.vim#L438.

File: doc/coc.txt (hunk 4 above)
Comment (fannheyward) [reply doc/coc.txt:2940]: https://github.com/neoclide/coc.nvim/pull/3355/commits/c5251c2b3418a881eef757e131a76898ceeb43c0

I've linked `CocSem` to `TS` hi group, the problem is we can't provide all needed hi group. The first part is synced with https://github.com/microsoft/vscode-languageserver-node/blob/main/types/src/main.ts#L3480-L3531, the second part is came from nvim-treesitter's hi group, and LS will provide some more token/modifiers.

**Comment by chemzqm** [comment]
@jackhub No need to hack the code any more, composed highlight group would work, checkout `:h CocSem`

**Review by chemzqm — APPROVED** [review]

**Comment by jackhub** [comment]
> @jackhub No need to hack the code any more, composed highlight group would work, checkout `:h CocSem`

Thanks for your great works.

**Comment by daquexian** [comment]
why semanticTokens.filetypes is empty by default? It should be `['*']` that aligns with the old coc.preferences.semanticTokensHighlights.

**Comment by chemzqm** [comment]
Because we don't want to surprise users with changed highlights.

**Comment by damnskippy** [comment]
Does the user have to explicitly call `CocAction('semanticHighlight')` to get the highlights, or does coc.nvim automatically make this call if `semanticTokens.filetypes`  is enabled for the current `&filetype` in vim.
In vim I see semantic highlighting working after calling `CocAction('semanticHighlight')`, but not automatically, so wanted to confirm if this is expected behavior.  Thanks.

**Comment by chemzqm** [comment]
> does coc.nvim automatically make this call if semanticTokens.filetypes is enabled for the current &filetype in vim.

It should be automatic.

**Comment by damnskippy** [comment]
Thanks for confirming, I'll look into my environment to see what's going on.
